### PR TITLE
feat(pagerduty): translate alert to externalid

### PR DESCRIPTION
**Pull request by georgettica**
- use mapstructure to convert between alert.Body to struct
   - use gopkg.in/yaml to convert the 'notes' field to struct

### CODE REVIEW

**Comment by georgettica** [comment]
related to OSD-10081

**Comment by georgettica** [comment]
look at OSD-10158

File: pkg/pagerduty/pagerduty.go [hunk 1]
```diff
@@ -126,6 +128,50 @@ func (p PagerDuty) AddNote(incidentID string, noteContent string) error {
...
+
+	var err error
+	externalBody := externalCHGMAlertBody{}
+
+	err = mapstructure.Decode(data, &externalBody)
+	if err != nil {
+		return "", fmt.Errorf("cannot marshal externalCHGMAlertBody: %w", err)
+	}
+
+	internalRawData := externalBody.Details.Notes
```
Comment (ramonbutter) [inline pkg/pagerduty/pagerduty.go:143]: All in all, this looks ok to me. However, I am generally not sure about the definition of these structures for unmarshalling as there are just used in one place.
I would maybe just go the way over json/yaml unstructured here. But its probably just taste, so feel free to go ahead anyway: 

```
details, ok := data["details"].((map[string]interface{}))
if !ok{
    return nil, UnexpectedFormatErr{field="details", data=data}
}
notes := make(map[string]interface{})
err := yaml.Unmarshal([]byte(details), &notes)
if !err{
    return nil, UnexpectedFormatErr{field="Notes", data=details}
}
clusterid, ok :=  notes["ClusterId"]
if err != nil {
    return nil, UnexpectedFormatErr{field="ClusterId", data=notes}
}
return clusterid, nil

```

File: pkg/pagerduty/pagerduty.go [hunk 2]
```diff
@@ -126,6 +128,50 @@ func (p PagerDuty) AddNote(incidentID string, noteContent string) error {
...
+
+	return internalBody.ClusterID, nil
+}
+
+// externalCHGMAlertBody is a struct for manipulating CHGM body until the note object
+type externalCHGMAlertBody struct {
+	// Details are the alert details
+	Details struct {
+		// Notes are the user given notes on the cluster
+		Notes string `mapstructure:"notes"`
```
Comment (ramonbutter) [inline pkg/pagerduty/pagerduty.go:165]: this type is exported by the external lib?

Comment (georgettica) [reply pkg/pagerduty/pagerduty.go:165]: yeah, the `mapstructure` helps the external lib comprehend what is going on inside the struct

File: pkg/pagerduty/pagerduty.go (hunk 1 above)
Comment (georgettica) [reply pkg/pagerduty/pagerduty.go:143]: I agree they are used only in one place, bit as @rafael-azevedo mentioned and I support that, marshalling to map[string]interface{} is bad

Comment (ramonbutter) [reply pkg/pagerduty/pagerduty.go:143]: why is it bad? Isn't a standard format to map a string key to different values:
 
`key : value with type of {string, int, float, map[string]interface{}}`

Comment (georgettica) [reply pkg/pagerduty/pagerduty.go:143]: not really. if I can get something more concrete here I will choose to do so. this is why I am willing to push the third parties in. having types creates certainty.

anyways we should move this offline

File: pkg/pagerduty/pagerduty.go [hunk 3]
```diff
@@ -126,6 +128,77 @@ func (p PagerDuty) AddNote(incidentID string, noteContent string) error {
 	return nil
 }
 
+// ExtractExternalIDFromCGHMAlertBody extracts from an Alert body until an external ID
+// the function is a member function but doesn't use any of the types
+func (_ PagerDuty) ExtractExternalIDFromCGHMAlertBody(data map[string]interface{}) (string, error) {
```
Comment (rafael-azevedo) [inline pkg/pagerduty/pagerduty.go:133]: Can we come up with a Smaller function name ?

File: pkg/pagerduty/pagerduty.go [hunk 4]
```diff
@@ -126,6 +128,77 @@ func (p PagerDuty) AddNote(incidentID string, noteContent string) error {
...
+	}
+
+	notes, ok := notesInterface.(string)
+	if !ok {
+		return externalCHGMAlertBody{}, AlertBodyExternalCastErr{FailedProperty: ".details.notes", ExpectedType: "string", ActualType: reflect.TypeOf(notes).String()}
+	}
+
+	res := externalCHGMAlertBody{}
+	res.Details.Notes = notes
+	return res, nil
```
Comment (rafael-azevedo) [inline pkg/pagerduty/pagerduty.go:184]: cleaner here if you have
`return externalCHGMALertBody{details: notes}, nil `

Comment (georgettica) [reply pkg/pagerduty/pagerduty.go:184]: Sure

File: pkg/pagerduty/pagerduty.go (hunk 3 above)
Comment (shibumi) [reply pkg/pagerduty/pagerduty.go:133]: What about ExtractExternalID? Or ExtractExternalCGHMAID? If we look on the receiver of the function, I think it becomes clear that this function is about alert bodys, but I never worked with PagerDuty. You might come up with a better idea :)

Comment (georgettica) [reply pkg/pagerduty/pagerduty.go:133]: nope, that's a good idea, naming is hard :/

File: pkg/pagerduty/pagerduty.go (hunk 4 above)
Comment (georgettica) [reply pkg/pagerduty/pagerduty.go:184]: i had an anonymous struct in the `Details` field,  which made it hard to create the stuct there. so I removed the whole struct as I only need to extract the "notes" field

File: pkg/pagerduty/pagerduty.go [hunk 5]
```diff
@@ -126,6 +128,77 @@ func (p PagerDuty) AddNote(incidentID string, noteContent string) error {
...
+		return "", AlertBodyDoesNotHaveNotesFieldErr{}
+	}
+
+	internalBody := internalCHGMAlertBody{}
+	err = yaml.Unmarshal([]byte(internalRawData), &internalBody)
+	if err != nil {
+		return "", NotesParseErr{Err: err}
+	}
+
+	// TODO: verify if no ClusterID is missing (doesn't break the test case, but should be noted
```
[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/pagerduty/pagerduty.go:154]: no, I think I'll keep it as this TODO is for the case we have another parser and it arrives here.

@rafael-azevedo noted that this won't happen as if this case would occur this function should fail but I wanted to note it somewhere

[the remark this answers is not in the document]
Comment (shibumi) [reply pkg/pagerduty/pagerduty.go:154]: @georgettica Does it make sense to open a github issue for this todo? If yes, I can open one.

[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/pagerduty/pagerduty.go]: It doesn't really, but maybe we should track it with out internal ticketing
If you want I can show you the ropes there

[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/pagerduty/pagerduty.go]: added an error, so it's fine for now

[the remark this answers is not in the document]
Comment (shibumi) [reply pkg/pagerduty/pagerduty.go]: lgtm

File: pkg/pagerduty/pagerduty.go [hunk 6]
```diff
@@ -126,6 +128,76 @@ func (p PagerDuty) AddNote(incidentID string, noteContent string) error {
...
+	var ok bool
+	_, ok = body["details"]
+	if !ok {
+		return "", AlertBodyExternalParseErr{FailedProperty: ".details"}
+	}
+
+	details, ok := body["details"].(map[string]interface{})
+	if !ok {
+		err := AlertBodyExternalCastErr{
+			FailedProperty: ".details",
```
Comment (ramonbutter) [inline pkg/pagerduty/pagerduty.go:170]: What's about adding the body here as a parameter? When you print the error, you immediately got the input that has caused this exception

**Review by shibumi — APPROVED** [review]
/lgtm

**Comment by georgettica** [comment]
@shibumi 
It's fir a forward slash In the from `/lgtm`

File: pkg/pagerduty/pagerduty.go (hunk 6 above)
Comment (georgettica) [reply pkg/pagerduty/pagerduty.go]: done

**Comment by georgettica** [comment]
@shibumi editing is not captured by the bot, so you need to write it correctly the first time

(or file an issue somewhere, not sure where this should go)

**Pull request merged**